feat: log SkyPilot queue latency at job start

teodorionita opened the pull request:
# feat: log SkyPilot queue latency to WandB

## Summary
Track how long SkyPilot jobs wait in the queue before starting. This helps identify cloud provisioning delays.

## What's New
- **`skypilot_latency.py`**: Standalone script that calculates queue time from SkyPilot's task ID
- **WandB metric**: Logs `skypilot/queue_latency_s` to your training runs
- **Zero impact**: Runs independently, doesn't touch training code

## How It Works
1. SkyPilot sets `SKYPILOT_TASK_ID` with a timestamp when jobs are submitted
2. Our script calculates the time difference when the job actually starts
3. Logs the latency to WandB (if configured) and stdout

## Testing
Simulate a 30-second queue delay
```bash
export SKYPILOT_TASK_ID="sky-$(python3 -c "import datetime; print((datetime.datetime.utcnow() - datetime.timedelta(seconds=30)).strftime('%Y-%m-%d-%H-%M-%S-%f'))")_test_123"
export METTA_RUN_ID="test_$(date +%s)"
export WANDB_PROJECT="metta"
python common/src/metta/common/util/skypilot_latency.py
```

Output: `SkyPilot queue latency: 30.0 s`

Unit tests pass: `pytest common/tests/util/test_skypilot_latency.py`

## Files Changed
- `common/src/metta/common/util/skypilot_latency.py` - new latency calculator
- `common/tests/util/test_skypilot_latency.py` - tests
- `devops/skypilot/config/sk_train.yaml` - added script call

teodorionita commented:
@rjwalters PR ready for review for option 3! :D

teodorionita commented on common/src/metta/common/util/skypilot_latency.py:
```diff
@@ -0,0 +1,37 @@
+    sky-YYYY-MM-DD-HH-MM-SS-ffffff_<cluster>_<n>
+"""
+
+from __future__ import annotations
```
agreed, o3/opus artifact from earlier python versions :D I left it there but will remove

teodorionita commented on tools/train.py:
```diff
@@ -68,8 +69,14 @@ def main(cfg: ListConfig | DictConfig) -> int:
+        os.environ["SKYPILOT_QUEUE_LATENCY_S"] = str(latency_sec)
+        logger.info(f"SkyPilot queue latency: {latency_sec:.1f} s")
+    else:
+        logger.info("SkyPilot queue latency: N/A (not a SkyPilot job)")
```
let me know when you have a chance to test!

rjwalters reviewed (changes requested):
We want this to be a skypilot specific features, so I think it would be better to add a main() to the skypilot_latency.py file and a PEP723 header (similar to .github/scripts) that accepts ENV for run name and credentials. Then we can call this script from the skypilot bash launcher to keep things encapsulated and avoid changes in the main `train.py` entrypoint

teodorionita commented:
> We want this to be a skypilot specific features, so I think it would be better to add a main() to the skypilot_latency.py file and a PEP723 header (similar to .github/scripts) that accepts ENV for run name and credentials. Then we can call this script from the skypilot bash launcher to keep things encapsulated and avoid changes in the main `train.py` entrypoint

should I also move the `skypilot_latency.py` script somewhere else as we won't use it as a util import anymore?

rjwalters commented:
I think we should still attach this data to the wandb log -- but from `devops/skypilot/launch.py` rather than from `train.py`

teodorionita commented:
@rjwalters I've updated the implementation, the wandb logging now happens from the skypilot side! 

Since launch.py runs locally (before the job is queued), it doesn't have access to SKYPILOT_TASK_ID. so instead I did the following:

1. Added the latency calculation to `sk_train.yaml` run script (where SKYPILOT_TASK_ID exists)
2. The standalone skypilot_latency.py script handles both calculation and WandB logging
3. No changes to train.py or launch.py

This way the latency is calculated on the remote machine where it actually makes sense, and gets logged to wandb using the METTA_RUN_ID from the environment.

Let me know if this works and when you can test with an actual skypilot run!

rjwalters commented:
https://wandb.ai/metta-research/metta/groups/b.rwalters.0702.simple.pr1236.0/workspace

I don't think it is working yet.. I'll poke at it

rjwalters commented:
![image](https://github.com/user-attachments/assets/6dc2e19f-e686-4ca4-bdcd-0996d8aaf390)

rjwalters reviewed (commented):
well.. after some effort this is working... I think the main problems were:
1. trying to run the file as a module (python -m) didn't work, so changed to python (path)
2. logging shadowing issues (I'm surprised these didn't show up before)

I think this will need some clean up before we merge it... we should think about whether there is any point in keeping the log_wandb stuff (I added this for testing purposes) 

We can probably also simplify sk_train a bit

@teodorionita can you take a turn on it next?

teodorionita commented:
Thanks for testing it and getting it to work!
re the logging not showing up before, lol it's great we spotted it 

I think we should keep the logging for wandb and the additional logs in `skypilot_latency.py`, no point in not having them! 

I've cleaned the `sk_train.yaml` and adjusted the tests for skypilot latency to replicate the actual docker command (added some additional latency for the subprocess) and I have updated the PR description

I've also tested the wandb logging with some exported variables:

```bash
export SKYPILOT_TASK_ID="sky-$(date -u -v-30S '+%Y-%m-%d-%H-%M-%S')-123456_test_123"
export METTA_RUN_ID="test_$(date +%s)"
python common/src/metta/common/util/skypilot_latency.py
```

the only thing remaining I think is one more test for `sk_learn.yaml` in your cloud env before merging? 

<img width="520" alt="Screenshot 2025-07-03 at 11 52 46" src="https://github.com/user-attachments/assets/2c25ebb4-d501-4f13-9fa4-2170761d2959" />
1

rjwalters commented:
lgtm

![image](https://github.com/user-attachments/assets/7b6a99ac-3769-4774-b730-50403251fe06)
